File: 차봄(믿고 맡겨조)/backend/src/main/java/kr/co/himedia/config/FirebaseConfig.java
package kr.co.himedia.config;

import com.google.auth.oauth2.GoogleCredentials;
import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import jakarta.annotation.PostConstruct;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Configuration;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

@Slf4j
@Configuration
public class FirebaseConfig {

    @Value("${app.firebase.config-path}")
    private String firebaseConfigPath;

    @PostConstruct
    public void initialize() {
        try {
            if (FirebaseApp.getApps().isEmpty()) {
                InputStream serviceAccountStream = null;

                // 1. Try ClassPath (Standard Spring Boot)
                try {
                    org.springframework.core.io.ClassPathResource resource = new org.springframework.core.io.ClassPathResource(
                            firebaseConfigPath);
                    if (resource.exists()) {
                        serviceAccountStream = resource.getInputStream();
                        log.info("Loaded Firebase config from ClassPath: {}", firebaseConfigPath);
                    }
                } catch (Exception e) {
                    log.warn("Failed to load from ClassPath: {}", e.getMessage());
                }

                // 2. Try filesystem (src/main/resources - Local Dev)
                if (serviceAccountStream == null) {
                    try {
                        Path path = Paths.get("src/main/resources", firebaseConfigPath);
                        if (Files.exists(path)) {
                            serviceAccountStream = Files.newInputStream(path);
                            log.info("Loaded Firebase config from filesystem (dev): {}", path);
                        }
                    } catch (Exception e) {
                        log.warn("Failed to load from filesystem (dev): {}", e.getMessage());
                    }
                }

                // 3. Try filesystem (Root - Fallback)
                if (serviceAccountStream == null) {
                    try {
                        Path path = Paths.get(firebaseConfigPath);
                        if (Files.exists(path)) {
                            serviceAccountStream = Files.newInputStream(path);
                            log.info("Loaded Firebase config from filesystem (root): {}", path);
                        }
                    } catch (Exception e) {
                        log.warn("Failed to load from filesystem (root): {}", e.getMessage());
                    }
                }

                if (serviceAccountStream == null) {
                    log.error(
                            "Firebase config file not found. Checked ClassPath, src/main/resources, and root. Path: {}",
                            firebaseConfigPath);
                    return;
                }

                FirebaseOptions options = FirebaseOptions.builder()
                        .setCredentials(GoogleCredentials.fromStream(serviceAccountStream))
                        .build();

                FirebaseApp.initializeApp(options);
                log.info("Firebase application has been initialized");
            }
        } catch (Exception e) {
            log.error("Failed to initialize Firebase (App will continue without Firebase): {}", e.getMessage());
        }
    }
}

File: Bask-EAT((주)식회사)/Bask-EAT-Service-Backend/src/main/kotlin/com/bask_eat/business_backend/config/FirebaseConfig.kt
package com.bask_eat.business_backend.config

import com.google.auth.oauth2.GoogleCredentials
import com.google.cloud.firestore.Firestore
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.cloud.FirestoreClient
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import jakarta.annotation.PostConstruct

@Configuration
class FirebaseConfig {

  @Value("\${firebase.config.file-path}")
  private lateinit var firebaseConfigPath: String

  @PostConstruct
  fun initializeFirebase() {
    try {
      val cl = this::class.java.classLoader
      var serviceStream = cl.getResourceAsStream(firebaseConfigPath)
      if (serviceStream == null) {
        // fallback to direct file path
        val file = java.io.File(firebaseConfigPath)
        if (file.exists()) {
          serviceStream = file.inputStream()
          logger.info("Loaded Firebase credentials from file path: {}", file.absolutePath)
        } else {
          throw IllegalStateException("Firebase credentials not found at: $firebaseConfigPath (classpath and file path)")
        }
      } else {
        logger.info("Loaded Firebase credentials from classpath: {}", firebaseConfigPath)
      }

      val options = FirebaseOptions.builder()
        .setCredentials(GoogleCredentials.fromStream(serviceStream))
        .build()

      if (FirebaseApp.getApps().isEmpty()) {
        FirebaseApp.initializeApp(options)
        logger.info("Firebase DEFAULT app initialized")
      }
    } catch (e: Exception) {
      logger.error("Firebase 초기화 실패", e)
      throw RuntimeException("Firebase 초기화에 실패했습니다.", e)
    }
  }

  @Bean
  fun firestore(): Firestore {
    return FirestoreClient.getFirestore()
  }

  @Bean
  fun firebaseAuth(): FirebaseAuth {
    return FirebaseAuth.getInstance()
  }

  companion object {
    private val logger = LoggerFactory.getLogger(FirebaseConfig::class.java)
  }
}
